docs/config.md
Index.py 内置的配置类 `indexpy.config.Config` 是一个单例类，你可以在任何地方使用 `Config()`，它们都将返回同一个对象。

所有配置都是大小写无关的，但推荐在程序中使用大写。

在 Index 启动时，它将自动从环境变量与项目根目录下 index.yaml / index.config 里读取配置。

!!! notice
    在 Index 运行之后更改配置文件或者环境变量并不会触发 Index 的热更新，你只能通过重启来使用新配置启动 Index。

## 环境变量

Index 在启动时将从环境变量里读取 `INDEX_DEBUG` 和 `INDEX_ENV` 两个值.

`INDEX_DEBUG` 的值为 `True` 或者 `on` 则 `DEBUG` 为真，其他任何值都是假。

`INDEX_ENV` 的值可以是任何字符串，它对应 `ENV`。

!!! tip
    环境变量在读取配置文件之后读取，这意味你可以使用环境变量的配置来覆盖配置文件里的配置。

## 配置文件

下面是一个配置样例：

```yaml
# overwrite default value to this program
log_level: "info"
allow_underline: true,
allowed_hosts: ["localhost"]
cors_allow_origins: ["*"]
cors_allow_methods: ["GET"]
cors_allow_credentials: false

# use in development
dev:
    "host": "0.0.0.0"
    "debug": true

# use in production
pro:
    "host": "127.0.0.1"
    "port": "41900"
    "log_level": "warning"
    "hotreload": false
```

## 所有可用的配置

### ENV

**默认值:** `"dev"`

`env` 是一个十分重要的配置，它允许自动使用对应环境下的配置。

### DEBUG

**默认值:** `False`

我不认为这个配置需要解释什么。

在环境变量里 `INDEX_DEBUG` 为 `"on"` 或者 `"True"` 时，`DEBUG` 为真。

### LOG_LEVEL

**默认值:** `"info"`

`log_level` 有五个可用值, 下面是它与 `logging` 的等级对应表

log_level   | logging
---         | ---
"critical"  | logging.CRITICAL
"error"     | logging.ERROR
"warning"   | logging.WARNING
"info"      | logging.INFO
"debug"     | logging.DEBUG

### HOST

**默认值:** `"127.0.0.1"`

`host` 指定 Index 监听的地址。

### PORT

**默认值:** `4190`

`port` 指定 Index 监听的端口。

### ALLOWED_HOSTS

**默认值:** `["*"]`

`allowed_hosts` 用于指定 Index 允许被访问的 HOST。

一些例子:

  1. `["*"]`

    允许所有的 HOST 访问

  2. `["example.com", "*example.com"]`

    允许 example.com 以及 example.com 的子域名访问。

  3. `["example.com", "test.com"]`

    允许 example.com 与 test.com 的访问。

### FORCE_SSL

**默认值:** `False`

`force_ssl` 允许 HTTP/WS 强制跳转到 HTTPS/WSS。

### AUTORELOAD

**默认值:** `False`

`autoreload` 为真时，使用 `index-cli serve` 或 `index-cli gunicorn start` 时，将会监听当前的项目修改，自动重启服务。

### HOTRELOAD

**默认值:** `True`

`hotreload` 为真时，将允许 Index 提供真正的热重载功能。

!!! notice
    当 `AUTORELOAD` 为真时，此选项无效。

### ALLOW_UNDERLINE

**默认值:** `False`

由于 Python 的 module 只允许字母、数字与下划线，但 URI 中出现出现下划线是被谷歌不推荐的，所以有了这个配置选项。

当 `allow_underline` 为假时，如果 _ 存在于 URI 中，它将会被自动替换成 - 并且 301 跳转过去。

### CORS_ALLOW_ORIGINS

**默认值:** `()`

在跨域请求中允许的 ORIGIN 列表。

例如： `['https://example.org', 'https://www.example.org']`

你可以使用 `['*']` 来允许所有的 ORIGIN 值。

### CORS_ALLOW_ORIGIN_REGEX

**默认值:** `None`

需要是一个正则表达式字符串，它将用于匹配在跨域请求中允许的 ORIGIN 。

例如：`'https://.*\.example\.org'`

### CORS_ALLOW_METHODS

**默认值:** `("GET",)`

在跨域请求中允许的 HTTP 请求方法列表。

你可以使用 `['*']` 来允许所有的请求方法。

### CORS_ALLOW_HEADERS

**默认值:** `()`

在跨域请求中允许被使用的 header 列表。

你可以使用 `['*']` 来允许所有的 header。`Accept`、`Accept-Language`、`Content-Language` 和 `Content-Type` 在跨域请求中是默认允许的。

### CORS_ALLOW_CREDENTIALS

**默认值:** `False`

允许在跨域请求中携带 cookies。

### CORS_EXPOSE_HEADERS

**默认值:** `()`

指定可供浏览器访问的任何响应头。

### CORS_MAX_AGE

**默认值:** `600`

设置浏览器缓存 CORS 响应的最长时间（单位：秒）。

## 什么是配置隔离？

同一个项目，不同环境下的部分配置可能不同。Index 内置的配置允许使用 `ENV` 来指定当前使用的配置环境。

看下面这个配置文件

```yaml
# overwrite default value to this program
log_level: "info"
allow_underline: true,
allowed_hosts: ["localhost"]
cors_allow_origins: ["*"]
cors_allow_methods: ["GET"]
cors_allow_credentials: false

# use in development
dev:
    "host": "0.0.0.0"
    "debug": true

# use in production
pro:
    "host": "127.0.0.1"
    "port": "41900"
    "log_level": "warning"
    "hotreload": false
```

当 `ENV` 的值为 `"dev"` 时，`Config().DEBUG` 会为 `True`。

当你指定 `ENV` 的值为 `"pro"` 时, 在使用 `Config().DEBUG` 的时候，它将先从 `"pro"` 中查找 `"debug"`（不分大小写）。当没有找到时，继续向上查找。然而这份配置文件并没有在根配置中指定 `"debug"`，所以 `Config().DEBUG` 会使用默认值 `False`。

其他的配置同理。


indexpy/config.py
import os
import json
import typing
import logging

import yaml

from .utils import Singleton

__all__ = ["Config"]

LOG_LEVELS = {
    "critical": logging.CRITICAL,
    "error": logging.ERROR,
    "warning": logging.WARNING,
    "info": logging.INFO,
    "debug": logging.DEBUG,
}

here = os.getcwd()


class ConfigError(Exception):
    pass


class ConfigFileError(ConfigError):
    pass


class UpperDict:
    def __init__(self, data: dict):
        self.__dict: typing.Dict[str, typing.Any] = dict()

        for key in data.keys():
            self[key] = data[key]

    def __str__(self) -> str:
        indent = 4
        result = ["{"]

        def append(line):
            result.append(" " * indent + line)

        for key, value in self.__dict.items():
            if isinstance(value, UpperDict):
                append(f"{key}: {{")
                for line in str(value).splitlines()[1:-1]:
                    append(f"{line}")
                append("}")
            else:
                if isinstance(value, str):
                    append(f"{key}: '{value}',")
                else:
                    append(f"{key}: {value},")

        result.append("}")

        return "\n".join(result)

    def __setitem__(self, key: str, value: typing.Any) -> None:
        key = key.upper()

        if isinstance(value, dict):
            if key in self.__dict.keys():
                for k, v in value.items():
                    self.__dict[key][k.upper()] = v
            else:
                self.__dict[key] = UpperDict(value)
        else:
            self.__dict[key] = value

    def __getitem__(self, key: str) -> typing.Any:
        return self.__dict[key.upper()]

    def __delitem__(self, key: str) -> None:
        del self.__dict[key.upper()]

    def update(self, data: dict) -> None:
        for key in data.keys():
            self[key] = data[key]

    def get(self, key: str, default=None) -> typing.Any:
        try:
            return self[key]
        except KeyError:
            return default


class Config(UpperDict, metaclass=Singleton):
    def __init__(self) -> None:
        super().__init__({})
        self.setdefault()
        # read config from file
        self.import_from_file()
        # read config from environ
        self.import_from_environ()

    def import_from_file(self) -> None:
        filename = None

        for _filename in ["index.json", "index.yaml", "index.yml"]:
            if os.path.exists(os.path.normpath(_filename)):
                if filename is not None:
                    raise ConfigFileError(
                        f"`{filename}` and `{_filename}` cannot be used at the same project."
                    )
                filename = _filename

        if filename is None:
            return

        with open(filename, "rb") as file:
            if filename.endswith(".json"):
                data = json.load(file)
            elif filename.endswith(".yaml") or filename.endswith(".yml"):
                data = yaml.safe_load(file)

        if not isinstance(data, dict):
            raise ConfigError(f"config must be a dictionary.")

        self.update(data)

    def import_from_environ(self) -> None:
        result: typing.Dict[str, typing.Any] = {}

        if os.environ.get("INDEX_DEBUG"):
            result["debug"] = os.environ.get("INDEX_DEBUG") in ("on", "True")

        if os.environ.get("INDEX_ENV"):
            result["env"] = os.environ.get("INDEX_ENV")

        self.update(result)

    def setdefault(self) -> None:
        """set default value"""

        self["env"] = "dev"
        self["debug"] = False
        self["host"] = "127.0.0.1"
        self["port"] = 4190
        self["log_level"] = "info"
        self["hotreload"] = True
        self["autoreload"] = False
        # url
        self["allow_underline"] = False
        # middleware
        self["force_ssl"] = False
        self["allowed_hosts"] = ["*"]
        self["cors_allow_origins"] = ()
        self["cors_allow_methods"] = ("GET",)
        self["cors_allow_headers"] = ()
        self["cors_allow_credentials"] = False
        self["cors_allow_origin_regex"] = None
        self["cors_expose_headers"] = ()
        self["cors_max_age"] = 600

    def __getattr__(self, name: str) -> typing.Any:
        value = self.get(name, ...)
        if value is ...:
            raise AttributeError(
                f"'{self.__class__.__name__}' object has no attribute '{name}'"
            )
        return value

    def __setattr__(self, name: str, value: typing.Any) -> None:
        if name == f"_UpperDict__dict":
            return super().__setattr__(name, value)
        raise ConfigError("Modifying the attribute value of Config is not allowed.")

    def __delattr__(self, name: str) -> None:
        raise ConfigError("Modifying the attribute value of Config is not allowed.")

    def __delitem__(self, key: str) -> None:
        raise ConfigError("Modifying the attribute value of Config is not allowed.")

    def __setitem__(self, key: str, value: typing.Any) -> None:
        key = key.upper()

        if key == "DEBUG":
            value = bool(value)
        elif key == "PORT":
            value = int(value)
        elif key == "ALLOWED_HOSTS":
            value = list(value)
            value.append("testserver")

        super().__setitem__(key, value)

    def get(self, key, default=None) -> typing.Any:
        env = super().get(self["env"], {})
        value = env.get(key, ...)
        if value is ...:
            value = super().get(key, default)
        return value
